src/bvals/cc/bvals_cc.cpp

```diff
@@ -133,7 +133,9 @@ int CellCenteredBoundaryVariable::LoadBoundaryBufferSameLevel(Real *buf,
   sk = (nb.ni.ox3 > 0) ? (pmb->ke - NGHOST + 1) : pmb->ks;
   ek = (nb.ni.ox3 < 0) ? (pmb->ks + NGHOST - 1) : pmb->ke;
   int p = 0;
-  BufferUtility::PackData(var_cc, buf, nl_, nu_, si, ei, sj, ej, sk, ek, p);
+
+  ParArray4D<Real> var_cc_ = var_cc.Get<4>(); // automatic template deduction fails
```

Comment:
Hmm this should not fail. I can look into it if you want. What concerns me is that if you explicitly set it to `ParArray4D` that assumes a layout. On the other hand, as we discussed on the call, if we change layouts we're dealing with more problems anyway.

Reply:
ParArray4D also uses the Layout specified by the LayoutWrapper so it should be agnostic to the layout (or at least in itself consistent with the other ParArrays).
Feel free to see why the automatic deduction fails.
I assume that it's related to the `device_view_t` and `host_view_t` defined in `parthenon_arrays.hpp`.

Reply:
> ParArray4D also uses the Layout specified by the LayoutWrapper so it should be agnostic to the layout (or at least in itself consistent with the other ParArrays).

The problem is that if you switch the layout for `ParArrayND`, then `var_cc.Get<4>()` might return an array with `LayoutTile`

Reply:
I see.
I'm starting to wonder if it would be preferable to template the majority of methods so that we can have specialization on `ParArrayND`s and `ParArray#D`s.
As far as I can tell all use cases we have in mind don't require sparse slices etc. so we could optimize the methods and kernels for a fixed layout (!= tiled).

Reply:
I think templating on array type is reasonable. Although it does mean you are less explicit about dimension. I don't think it's a huge deal either way at the moment, so long as we are cognizant of what we're doing.

Reply:
> Feel free to see why the automatic deduction fails.

I think I see what's going on.

 The type returned by `ParArrayND.Get<4>()` is:
```C++
 Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >
```
But `ParArray4D` is 
```C++
Kokkos::View<double*****, Kokkos::LayoutRight, Kokkos::HostSpace>
```
Fortunately, the two types are compatible. So you can set one equal to the other and do implicit casts. i.e., the following works:
```C++
void doNothingByValue(ParArray4D<Real> array) {}
ParArrayND<Real> ndarray;
auto b = ndarray.Get<4>();
doNothingByValue(b);
```
But C++ won't do implicit casts on references to variables, so *this* fails:
```C++
void doNothingByReference(ParArray4D<Real>& array) {}
ParArrayND<Real> ndarray;
auto b = ndarray.Get<4>();
doNothingByValue(b);
```
I started an issue to discuss this in more detail, #143. I don't think it's a huge deal, though. And your explicit typecasting seems like a fine solution for now.

Reply:
On `ParArray4D` vs `ParArrayND` - `var_cc` is a single variable now which is accessed and iterated over a 4D array. I prefer the hard typing so that the compiler handles errors. However, Changing the layout so that `ParArray4D` ends up being a `LayoutTile` will probably add overhead in the form of extra integer arithmetic and register usage. It shouldn't lead to strided access patterns since the iteration pattern will change with the layout.

Long term though, these boundary functions should merge variables in one kernel, so this variable will become a `VariablePack`

Reply:
Yes, those functions are very likely to see an update anyway as they are one of the performance issues in K-Athena.
I even expect we not only need merge Variables but actually fill all buffers in a single kernel simultaneously.

Reply:
Right. I wouldn't be too concerned about it now. As I describe in the issue, the hard typing should work, as should passing the view into functions by value instead of reference.

Reply:
The reason I dug in is that I was worried `ParArrayND` had a bug related to type inference, but I think it's all working as intended.